### DISCUSSION TITLE
LilyGO T-TWR support

### DISCUSSION THREAD

**Discussion by lyusupov** [erstec/APRS-ESP#29] [Ideas]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Hi, Ernest

Thank you for development of this great project !

It looks like LilyGO is about to release soon an ESP32-S3 DIY kit with SA868 radio and OLED display on board -

![sa868](https://user-images.githubusercontent.com/5849637/204091506-02b9d590-baaf-4c0c-bd0a-6e215fe934c6.jpg)

This is the GitHub page of the upcoming product: https://github.com/Xinyuan-LilyGO/T-TWR
and these are the docs: https://t-twr.readthedocs.io/en/latest/

Few notes:
1) it does not look like they have a GNSS module on-board ;
2) not sure if they have connected the MIC and SPEAKER pads of the SA868 to the ESP32-S3 ones.

However, the T-TWR product has a 2.54 mm pin header, so a user can add a GNSS option and wire MIC/SPK connections by him/herself.

My proposal to you is: 
please, consider to add support for the T-TWR kit - this will likely increase  your project audience considerably.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  Best regards,
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  Linar.

**Comment by erstec** [comment]
@lyusupov 

Hi Linar!

Thank you for such interesting information!
It is a good idea and I looked at documentation of course.
Board have almost all needed to make it work.

Regarding Implementation:
I don't practice programming of embedded systems without having embedded system on my bench. It is almost impossible to make working thing without debugging and adjusting to specific needs for every board. And I have no plans to do "reverse engineering" based on their examples as I can't be sure they works properly and, tbh, it is wrong ways to have finished device.

So, at first we need to wait till this product will become available to buy it and next (if I be able and can afford) will buy sample or two for real development. 
Or... Anyone who want to have support of APRS-ESP on this board can make effort and implement it.
I very appreciate everyone involvement to the project and waiting for any pull requests :) regardless "good" or "bad", regardless style and anything elese. All code can be adapted and merged.
Or, another way, let someone provide me sample of this LilyGo board and I will make all required things, including "motherboard" (see below) hardware and Gerbers.

Next thing is that I don't see any schematics provided by LilyGo to make 100% sure statements of "how it connected", but from what I see my comments are:
> 1. it does not look like they have a GNSS module on-board ;
As you told - it is solvable by connecting external via available header (same as on my board).
> 2. not sure if they have connected the MIC and SPEAKER pads of the SA868 to the ESP32-S3 ones.
Mic/Speaker connection is not done straight way. If you will look at my schematics it have matching circuits for both of them. And be sure, you can't get proper modulation and decoding without using them. So, in this case "motherboard" will be needed still.

And last, but not least question is price, let see how much will it cost.

Anyway, there is nothing wrong to add support of it. But only after LilyGo release this product 
It can be portable device, battery and/or USB powered, of course!

I highly appreciate information you provided and that you are interested in project and want ask you to let me know if Schematics and/or possibility of buying this board will became available and you will notice it. Drop me a message, please, in this case

BR,
Ernest.

**Comment by lyusupov** [comment]
&nbsp; &nbsp; Ernest,

don't too much  worry about schematics, samples and other stuff - LilyGO is making a lot of efforts to support DIY developers.
If they will qualify your project to be eligible for their Devs Support Program _(which I am sure you are)_  - they will supply all you need for free.

Please, wait few days. LilyGO or me will likely contact to you next week.

 &nbsp; &nbsp; Best regards,
 &nbsp; &nbsp; Linar.

**Comment by lyusupov** [comment]
Be aware that ESP32-**S3** has no built-in DACs:
https://www.esp32.com/viewtopic.php?t=22740

Software should use either:
- external hardware I2S DAC or
- [PDM_MONO_MODE](https://github.com/espressif/arduino-esp32/blob/master/libraries/I2S/src/I2S.h#L60) method ([Espressif](https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/i2s.html#pdm-tx-usage)) with external RC circuit

Current implementation of AFSK has support for :
* [external DAC](https://github.com/erstec/APRS-ESP/blob/master/lib/LibAPRS_ESP32/AFSK.cpp#L144) and
* [built-in DAC](https://github.com/erstec/APRS-ESP/blob/master/lib/LibAPRS_ESP32/AFSK.cpp#L165)

**Comment by lyusupov** [comment]
Schematic file of the T-TWR V1.3 is available for public at this location:
https://github.com/Xinyuan-LilyGO/T-TWR/tree/master/sch

T-TWR price range can be estimated by this listing:
https://www.aliexpress.com/item/1005005133900444.html

T-TWR V1.3 is not ready yet for the APRS feature.
However, there is a T-TWR version of V1.5 _(which is currently under development)_ that is expected to be "APRS ready", It has certain new features, such as:
- analog output to SA868
- analog input from SA868
- GNSS module

**Comment by lyusupov** [comment]
Few words about I2S PDM mode in ESP32-S3 that can be used to implement the "digital to a analog" conversion.

It was confirmed that this method works quite good -

## Hardware

![testbed](https://user-images.githubusercontent.com/5849637/219305668-7834e7e9-1171-439f-a80a-8650e2536e19.jpg)

LPF (Low Pass Filter) is a simple RC filter circuit with
R = 1.5 kOhn and C = 10 nF
 
## Software

Sketch PlayWAVFromFunction from ESP8266Audio library:
https://github.com/earlephilhower/ESP8266Audio

## Results


![ESP32_DAC](https://user-images.githubusercontent.com/5849637/219306011-37539fb1-fee5-4b68-9635-ea00e50344a4.jpg)
![ESP32_DAC_LPF](https://user-images.githubusercontent.com/5849637/219306015-e38920be-e069-47b8-a467-de5fe4a8f282.jpg)
![ESP32_PDM](https://user-images.githubusercontent.com/5849637/219306020-197a1994-6e64-44f5-971b-b9e8c6022f2c.jpg)
![ESP32_PDM_LPF](https://user-images.githubusercontent.com/5849637/219306025-b27df72d-2ce6-4fa9-9157-00bb8e462e64.jpg)
![ESP32S3_PDM](https://user-images.githubusercontent.com/5849637/219306026-3fcc782e-104a-44c3-bd84-2d7077e1b1cf.jpg)
![ESP32S3_PDM_LPF](https://user-images.githubusercontent.com/5849637/219306029-97186d84-f4db-4a56-a1a2-51117310345c.jpg)

## Summary

1) on ESP32 the PDM output with LPF works as good as it's  DAC mode does
2) PDM mode on ESP32-S3 works exactly the same way as it does on a regular ESP32

## Conclusion

ESP32-S3 does not have built-in DAC but the I2S PDM mode works quite good to produce basic analog signals.

**Reply by erstec** [reply]
Nice! Thnak you for information and testings!

**Comment by lyusupov** [comment]
FYI:

Rev. 2 of the LilyGO T-TWR has the following new features:
- receive of SA868 audio output signal by ESP32-S3
- ESP32-S3 can supply PDM signal to SA868 MIC input
- built-in GNSS receiver

Schematics: https://github.com/Xinyuan-LilyGO/T-TWR/blob/master/schematic/T-TWR-Plus_Rev2.0.pdf

Test snippets:
- [GNSS](https://github.com/Xinyuan-LilyGO/T-TWR/blob/master/examples/Rev2.0/GPS_Full_Example/GPS_Full_Example.ino)
- [Audio In](https://github.com/Xinyuan-LilyGO/T-TWR/blob/master/examples/Rev2.0/SA868_GetAudioRaw_Example/SA868_GetAudioRaw_Example.ino)
- [Audio Out](https://github.com/Xinyuan-LilyGO/T-TWR/blob/master/examples/Rev2.0/SA868_ESPSendAudio_Example/SA868_ESPSendAudio_Example.ino)

Exterior:

![](https://user-images.githubusercontent.com/5849637/233844863-ab5f9a9e-cb68-409c-96a8-d01c9809c564.jpg)

**Reply by erstec** [reply]
Is it already in stock?

**Reply by lyusupov** [reply]
Not yet,
AFAIK, V2 had experienced one more design & PCB update cycle.
If the upcoming batch will have no any critical issues - it may become released for sale to very first customers.

**Reply by erstec** [reply]
It would be nice to be in "very first customers" line, tbh :)

**Reply by lyusupov** [reply]
[**Lewis**](https://github.com/lewisxhe) is the LilyGO **engineer who made the design of V2**.
Try to contact with him to negotiate a free or reduced price board sample to you for your project development purpose.

**Reply by lyusupov** [reply]
> Is it already in stock?

https://www.lilygo.cc/products/t-twr-plus

**Comment by lyusupov** [comment]
**'Stock' T-TWR Plus** (ESP32-S3) built-in **I2S PDM** out over GPIO18 -> RC -> SA868  is doing good :

<br>

![image](https://github.com/erstec/APRS-ESP/assets/5849637/d66f8da0-a699-488f-9791-1eb933b247c5)

**Reply by erstec** [reply]
Nice! How about RX part?

**Reply by lyusupov** [reply]
it may require to backport this example
https://github.com/espressif/esp-idf/blob/release/v5.0/examples/peripherals/adc/continuous_read/README.md
from ESP-IDF V5.x to V4.4.X _(in use by Arduino Core 2.0.x)_

**Comment by lyusupov** [comment]
![](https://github.com/lyusupov/SoftRF/assets/5849637/382b1784-5717-4afb-be4f-b492408e588f)

<br>

![](https://github.com/lyusupov/SoftRF/assets/5849637/5af3f306-31f4-4458-9e3a-1febe48098b6)

<br>

![](https://github.com/lyusupov/SoftRF/assets/5849637/98b7e170-a429-4c61-bb0c-18569ceffc0b)

**Reply by erstec** [reply]
Perfect! Really waiting your contribution with support os S3 MCU, as I absolutely out of free time currently and can only make small fixes...

**Comment by lyusupov** [comment]
1) Lewis said that they will fix R22 value in the next batches, so we should probably wait few weeks for the current stocks to be sold out
2) "continuous DMA read" from IDF 5.x is hard to use with Arduino Core 2.0.x 
    It is worth to wait few months until the Core 3.x will be released. It will be based on IDF 5.1
3) in SoftRF, I currently use h/w timer based ADC poll @ 9600 SPS sampling rate that is available in the LibAPRS_ESP32 _(been developed by @nakhonthai)_ as a build option
4) you may pick some tips by making a 'diff' of my [AFSK.cpp](https://github.com/lyusupov/SoftRF/blob/master/software/firmware/source/libraries/LibAPRS_ESP32/AFSK.cpp) and [AFSK.h](https://github.com/lyusupov/SoftRF/blob/master/software/firmware/source/libraries/LibAPRS_ESP32/AFSK.h) files against your **APRS-ESP** ones

**Reply by erstec** [reply]
Thank you for suggestions, but as I told before, I don't have time (and will not in next month) for active development of this part.

Anyway, except R22 value V2.0 have no SQL line routed and I don't understand why, it saves A LOT of power when device running from battery. Also AXP2101 have not enough plane for heat dispassion, charging @1Amp rises it temperature (and all around) up to 90C+, so "safe" charging are only possible on @~600mA

About plans: Firstly I need to migrate to U8glib definitely, as Adafruit drivers of SH1106 are really slow and causes ESP watchdog sometimes. Then I need to have some working and usable OLED menu system, that I'm working right now on, as for me, personally, device are very hardly usable on-field without this. Configuring through Web interface are not convenient from mobile at all...
So if someone step-in and take care on ESP32S3 things - it would be nice, otherwise it will happen, but not in near future

**Comment by lyusupov** [comment]
Few more illustrations for APRS Tx and Rx operations with 'stock' LilyGO T-TWR Plus:

<br>

One screenshot of a video taken by my mobile phone -

![](https://github.com/erstec/APRS-ESP/assets/5849637/48c0b21b-31d7-4c9e-848a-38281cf352da)

<br>

**Full video** file is available on Google Drive:
https://drive.google.com/file/d/18a2PpIE4KQ6lYPcmPZgXnCYFSRYEqKz-/view

<br>

These are some screenshots of the SoftRF WebUI:

![twr_softrf_aprs](https://github.com/erstec/APRS-ESP/assets/5849637/1e238eae-a442-4132-85e2-512d0fbfa86f)

**Comment by lyusupov** [comment]
![](https://github.com/erstec/APRS-ESP/assets/5849637/71c95a60-dc17-4e6e-a88b-036e8aba7c1a)

**Comment by nakhonthai** [comment]
Coming soon for ESP32IGATE-TWR Plus.

![20231012_214338.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/6f29cfab-d53b-4bab-9321-fbf56bc57262)

#Web Services 

![SmartSelect_20231012_091645_Samsung Internet.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/7f3ba816-83eb-4132-9c85-b86dd4c164c8)

![SmartSelect_20231012_091702_Samsung Internet.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/2b5e7dd7-e54f-4693-8ff5-7561b9977605)

#Test Sigmadelta sound output 
![1697096457598.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/1ec238ff-fb82-4826-b9af-c175fe85a2c9)

![1697096447124.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/1045b4c4-e675-4271-a54d-45052081e20f)

![1697096451476.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/c0a7d330-d061-4333-8a01-cf884a5f2ced)

**Reply by erstec** [reply]
Very nice job, @nakhonthai!
Thing that you used PiStar web design also good choice!

**Comment by lyusupov** [comment]
@nakhonthai the links above are broken...

Is that it: https://www.facebook.com/groups/aprsth/posts/6370012186367283/ ?

**Reply by nakhonthai** [reply]
Video clip in the first test. 
I'm redesign developing open source, coming soon.

**Reply by lyusupov** [reply]
Thanks @nakhonthai ! 

Is it possible to fix the broken pictures above ?

<img width="432" alt="image" src="https://github.com/erstec/APRS-ESP/assets/5849637/f6c7dae2-0253-48d3-b847-b63aefbdc48b">

**Reply by nakhonthai** [reply]
OK,I edited.

**Reply by lyusupov** [reply]
@nakhonthai 
> OK,I edited.

Thank you for the update!

I have a couple of questions

## Is this a typo ?

![imag](https://github.com/erstec/APRS-ESP/assets/5849637/b4ab6457-ca67-421b-bbc8-e047953ef00f)

<br>

## Have you achieved this Rx range with [**T-TWR Plus 'stock' antenna**](https://github.com/lyusupov/SoftRF/wiki/Ham-Edition#antenna) ?

![](https://github.com/erstec/APRS-ESP/assets/5849637/f7db6984-c917-4213-9a23-9968771830e8)

**Reply by nakhonthai** [reply]
Oh...Thank for Herd to Heard.
IGATE mode  can receive from internet and  rf.
TWR  Antenna can receive from RF <1km ,testing use antenna on tower.

**Reply by lyusupov** [reply]
@nakhonthai 

> TWR Antenna can receive from RF <1km

I've managed to get 1+ km range with the 'stock' antennas on both sides.
Both devices were within LoS. It, probably, could be possible to get more range- but it is hard to find an LoS distance of 2+ km in my local area.
APRS packets were received all the time while audio signal _(with SA868 SQL=1 setting)_ was monitored with an appropriate quality.

I believe you've replaced wrong R22 value with 47K by yourself, right ?

![](https://user-images.githubusercontent.com/5849637/271892784-080acdf6-d67e-4f29-8cc9-8b7229490c39.png)

**Reply by erstec** [reply]
@lyusupov I replaced stock cheap esp32-s3 with proper esp32 and it's working right now as -10 node with stock antenna, placed near my window for 6.7 km, you can see it on aprs map.

**Reply by erstec** [reply]
Also it is unaccpetable to don't have proper squelch line from rf modem to esp32, as power consuption rises 3 times and 21700 battery last only for ~2h

**Reply by lyusupov** [reply]
@erstec thanks for the report!

I can receive a strong APRS signal & data from [R2AKT](https://aprsdirect.de/?center=55.9382,37.973&zoom=16) iGate @ 26 km distance from my location by:
1) trusty-old (20+ years) aviation Vertex / Yaesu VXA-700  (dual 'air and 2m') band radio that I use as a backup while flying aircrafts, and
2) this year Quansheng UV-K5(8) handheld radio

However, the T-TWR Plus radio can not receive the same signal with the same antenna and pick the APRS data from the source. It means that there is still an area for the SA868 sensitivity improvements...

**Reply by erstec** [reply]
SA modems have a lot of sensitivity problems, even Misrosat devices I own with built-in SA not perfors as well as external radio, sure. More, stock antennas I got with T-TWR Plus was in resonance on ~450MHz (I believe it is because of target PMR frequencies), in comparision stock from T-TWR V1 was much more well adjusted to HAM freq. Regarding this, it somehow works ) only "problem" for this day is ESP32-S3, I don't have time to even look at it.

**Reply by lyusupov** [reply]
@erstec 
> 21700 battery last only for ~2h ...

1) Measured **standby current** of the T-TWR Plus board when
* ESP32-S3 is active with Wi-Fi component in AP state
* GNSS is active (acquisition)
* SA868 is in Rx state with SQ=1 setting

**does not exceed 220-250 mA average**

2) Typical genuine LG or Samsung **21700 battery capacity is 4000-5000 mAh**.

Thus, if I would have "21700 battery last only for ~2h" issue - I would definitely inspect
1) whether or not the battery weight / capacity comply with the specs and
2) whether or not a particular TWR Plus board has hidden 'current leaks'.

**Reply by erstec** [reply]
It's all doesn't matter, as Lilygo should develop they boards properly (as they cost 2 or 3 times of Quensheng price and same as baofeng ht's). If RF modem have this signal it should be routed to MCU. Use this signal or not is purely developer/user decision, not development board manufacturer. It is development board purpose to have all connections and all possibilities.

**Reply by lyusupov** [reply]
@erstec Please, give me a reference on a mass product of a battery powered APRS tracker with iGate, GNSS, Wi-Fi and Bluetooth features at cost of below 65 USD.

**Reply by erstec** [reply]
First of all LilyGo don't provide end user all you mentioned, as well as case or any other enclosure, they sold hardware only, in other words - development board. Other "solutions" are usable out-of-the-box, nevermind.
Second is that it cost more than 65 USD, as you somehow missed shipping costs, VAT and customs fee. Final price is ~90USD for receiver in EU.
Third, which part of T-TWR rises price to 60+USD? RF Module? T-Beam have absolutely same features, except LoRa SX instead of SA RF modem, so, it all just because I see.
And last one, instead of do it proper way (SQL line, wrong value resistor) and place already supported MCU (which price is bigger, ofc, but already supported) they release someting.

If you like way they do hardware - it is your choice.

Discussion are closed. I will stay with my opinion regardles anything.

**Reply by lyusupov** [reply]
@erstec 
We, lazy consumers, all wish you good luck with your own hardware design explained on the APRS-ESP project front page !
We will be happy to purchase your APRS-ESP "development board" in the future when (if) it will be ready for sale at or below the  T-TWR Plus current price tag. 
We are pretty much sure that your price for the APRS-ESP will not exceed the net sum of all the components costs.

**Reply by nakhonthai** [reply]
> @nakhonthai 
> 
> > TWR Antenna can receive from RF <1km
> 
> I've managed to get 1+ km range with the 'stock' antennas on both sides.
> Both devices were within LoS. It, probably, could be possible to get more range- but it is hard to find an LoS distance of 2+ km in my local area.
> APRS packets were received all the time while audio signal _(with SA868 SQL=1 setting)_ was monitored with an appropriate quality.
> 
> I believe you've replaced wrong R22 value with 47K by yourself, right ?
> 
> ![](https://user-images.githubusercontent.com/5849637/271892784-080acdf6-d67e-4f29-8cc9-8b7229490c39.png)
> 

Not change R22.

![SmartSelect_20231013_215728_Electrodoc Pro.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/2482c670-555f-4a67-89c2-85629d963318)

Test send TX  by Yaesu VX8-DR and  point check AUDIO2ESP pin, SA868 set  volume=8 max,You can see top cycle at  2.56V and cycle 1.56Vp-p

![1697215474652.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/71b317b7-7ab1-457e-8759-3f0f3f99ddd5)



I use adc  attention 11db and auto dc offset code.
You can found code from ESP32IGate project.

**Reply by nakhonthai** [reply]
Antenna 5/8y height 15m and BPF receiver by TWR Plus.
![SmartSelect_20231013_220618_Samsung Internet.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/92fbe93c-b66b-4f53-b69d-f48a4bb3b3fb)

However, RF Module SR_FRS_1W better SA818/SA868. It to  use my product many more. 
![Screenshot_20220608-183500_Gallery.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/a4d4c497-b833-4906-962a-59c8b25ef5b0)

![Screenshot_20220405-134454_LINE.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/6514e1a6-5803-4646-acb2-45233298048e)

![20230718_100910.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/e7db30d2-6f1d-46e1-a77c-b6cd4d2eab98)

![20221221_175734.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/ad8ae1a3-4e47-40e4-8a2b-25ef107660a0)

![20221205_164547.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/d3a6debe-96ae-43cf-845d-4f3bc8b367fd)

**Comment by lyusupov** [comment]
@nakhonthai 

> Not change R22.
> I use adc attention 11db and auto dc offset code.

Use of R22 = 4.7K leads to **unwanted 11 dB attenuation** to meet ESP32-S3 ADC max. voltage specs:
https://docs.espressif.com/projects/esp-idf/en/v4.4.5/esp32s3/api-reference/peripherals/adc.html#adc-attenuation

![](https://github.com/erstec/APRS-ESP/assets/5849637/82b48e80-237c-4b00-92b1-e460da458b66)

I have two T-TWR Plus VHF boards. 
1) I've replaced 4.7K with 47K **on the 1st board only** and changed calibration & attenuation onto 0dB as follows:
https://github.com/lyusupov/SoftRF/blob/master/software/firmware/source/libraries/LibAPRS_ESP32/AFSK.cpp#L379-L402

```
  if (hw_info.model == SOFTRF_MODEL_HAM) {
    if (hw_info.revision == 20) {
      uint64_t mac = ESP.getEfuseMac();
      if (mac == 0x58f8ab188534ULL /* || mac == 0x7475ac188534ULL */) {
        adc1_config_channel_atten((adc1_channel_t) SPK_PIN, ADC_ATTEN_DB_0);  // Input 1.24Vp-p,Use R 47K-(10K//10K) divider input power 1.2Vref
      } else {
        /* work around wrong R22 value (should be 47K) issue on very first T-TWR Plus batches */
        adc1_config_channel_atten((adc1_channel_t) SPK_PIN, ADC_ATTEN_DB_11); // Input 3.3Vp-p,Use R 10K divider input power 3.3V
      }
    } else {
      /* T-TWR Plus V2.1 is expected to use 47 KOhm value for R22 */
      adc1_config_channel_atten((adc1_channel_t) SPK_PIN, ADC_ATTEN_DB_0);  // Input 1.24Vp-p,Use R 47K-(10K//10K) divider input power 1.2Vref
    }
```
  2) I keep using 2nd board with R22=4.7K and 11dB attenuation to be able to compare both boards  by an outdoor test.

**My outdoor test showed me that the 1st board has about 3X better APRS Rx range** than the 2nd one.
This complies to increase of 11 dB in sensitivity.

**Reply by nakhonthai** [reply]
SA868 audio signal hight level output >2Vp-p ,in circuit of TWR Plus voltage ref at 2.6V is OK.

SR_FRS or other  module audio signal low level output <0.5Vp-p, I designed voltage divider at  1.2V (R22=33K) or  ic ref TL432 is better. 

My TWR Plus use testing is model  SA868.

**Reply by lyusupov** [reply]
@nakhonthai 

> SA868 audio signal hight level output >2Vp-p ,in circuit of TWR Plus voltage ref at 2.6V is OK.

So your suggestion is to set AT+DMOSETVOLUME=8 to the maximum level _(while keeping R22 = 4.7K)_ to compensate 11 dB attenuation inside the ADC component, right ?

**Comment by lyusupov** [comment]
@nakhonthai 

![](https://user-images.githubusercontent.com/5849637/261082041-8dd62a9b-af90-4afb-8468-238ca0d6750e.png)

Yes, this is a well known issue with SA868 that was reported 
1) here https://www.egimoto.com/a/17cf8552e5016036
2) and there: https://github.com/OpenRTX/sa8x8-fw/issues/8#issuecomment-1680974932

However, both me and **Tech Minds** YouTube blog https://www.youtube.com/watch?v=IImZfDASYIw&t=47s found no substantial harmonics  in the Tx spectrum of the T-TWR Plus board.

Have you noticed by yourself  any critical issues with the fact that the LC filter is missing in the SA868 PCB ?

**Reply by nakhonthai** [reply]
SA868 has bypass LPF , because 2W high power RF output. 

Receive has interference,When using an external antenna  BPF required.

Harmonics output of  TWR  Plus 
![TWR_HARMONICS.png](https://github.com/erstec/APRS-ESP/assets/16043758/cb4c1a5d-6032-430b-8ace-6f42f600a739)

**Reply by lyusupov** [reply]
@nakhonthai  thank you for the board spectrum picture !

> Receive has interference, When using an external antenna BPF required.

This is a common issue for most of inexpensive radios on the market that uses DCR ICs, such as RDA1846s inside the SA868. We can do nothing about it.  Superheterodyne ones are not considered in this price range.

In August this year [**Lewis**](https://github.com/lewisxhe) have added an option to use an LC filter in upcoming revision 2.1 of the T-TWR Plus:

![image](https://github.com/erstec/APRS-ESP/assets/5849637/d507bcb1-5146-4da7-9230-8b56819e9a8c)

**Reply by nakhonthai** [reply]
Easy to use add LFCN-160

![SmartSelect_20231014_142614_OneDrive.jpg](https://github.com/erstec/APRS-ESP/assets/16043758/c67dda2f-5c2d-463c-9dbb-a9e4cc00495e)

**Reply by lyusupov** [reply]
@nakhonthai 
> LFCN-160

 4+ USD per 1 piece ? It is even more than onboard ESP32-S3-WROOM-1 N16R8 price...
I personally doubt that LilyGO will follow this way...

Stuff like that would make sense when they will aim for FCC/CE compliance for the T-TWR. I doubt that it is a short-term goal.

**Comment by lyusupov** [comment]
[OpenRTX project](https://openrtx.org/#/hardware/ttwrplus) has made an UF2 firmware for LilyGO T-TWR Plus available in a [nightly build](https://files.openrtx.org/nightly/) as a part of this [Feature Request](https://github.com/OpenRTX/OpenRTX/issues/197).

**Comment by lyusupov** [comment]
Video of LilyGO T-TWR Plus **V2.1**: https://twitter.com/lilygo9/status/1752865846727446786
Schematic file: https://github.com/Xinyuan-LilyGO/T-TWR/blob/master/schematic/T-TWR-Plus_Rev2.1.pdf

![](https://raw.githubusercontent.com/lyusupov/SoftRF/master/documents/images/ham-29.jpg)

<br>
<br>

![](https://raw.githubusercontent.com/lyusupov/SoftRF/master/documents/images/ham-28.jpg)

**Comment by lyusupov** [comment]
[**T-TWR V2.1 is in the Store !**](https://www.lilygo.cc/products/t-twr-rev2-1)

![](https://www.lilygo.cc/cdn/shop/files/H709-T-TWRREV2_5.jpg)
